### PR TITLE
feat: add graph ascii representation

### LINKED ISSUE #100: Viz/graph of tasks
`whiz --graph` should output an ascii visualisation of the tasks and their dependency. Format is free but should support large graphs.

### PR DESCRIPTION
**Pull request by Darioazzali**
Added graphical ascii representation fix #100.
There is the possibility too to be visualized with [box-drawing character](https://en.wikipedia.org/wiki/Box-drawing_character).
The task resolved from the `run` function in main are used to build the node edge structure.
Then is visualized with ratatui as independent and dependent task, the second one as an ascii graph. 
The keys `k` `j` `l` and `h` permits the graphic navigation through scroll, as well as the arrows.
Would be grate disallow scrolling at the tasks graph boundary.
If you like the solution I can do it as well.
For drawing the edge and nodes I've used the [termgraph crate](https://docs.rs/termgraph/latest/termgraph/).

### CODE REVIEW

File: src/graph/graph_task.rs [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+            Graph::dependecies_lists_to_tuple_nodes(&two.depends_on, 2, &dependent_dictionary);
+
+        let dependencies_for_three =
+            Graph::dependecies_lists_to_tuple_nodes(&three.depends_on, 3, &dependent_dictionary);
+
+        assert_eq!(dependencies_for_two, Some(vec![(1, 2)]));
+        assert_eq!(dependencies_for_three, Some(vec![(1, 3), (2, 3)]));
+    }
+
+    // #[test] fn big_list_dep_list_to_nodes() {
```
Comment (zifeo) [inline src/graph/graph_task.rs:252]: rm?

**Comment by zifeo** [comment]
@Darioazzali Nice contribution, a little suggestion.

**Comment by zifeo** [comment]
@Darioazzali Nice contribution, a little suggestion. Can you also check the errors in the tests?

**Comment by Darioazzali** [comment]
@zifeo sorry I've run `cargo build` and everything was compiling well, didn't check for the test sorry for that (in the dev branch that  I've made I was using a test parsing from a big yaml and I was thinking if i should put it).

For the subcommand feature I've a couple of points to ask:
1. I've noticed in the main branch that if your run simply `whiz help` or `whiz help version` it outputs an error:
error
As you can see from the first line.
This is due to 
`    let args = Args::try_parse()?;`
https://github.com/zifeo/whiz/blob/081010c6371f069dbab880f48162ab1fa22b846a/src/main.rs#L59C24-L59C24

The error returned form the function is not kind of error, it's an error that Claps "throw" Clap [`DisplayHelp`](https://docs.rs/clap/latest/clap/error/enum.ErrorKind.html#variant.DisplayHelp).
I think a `Args::parse()` will suffice, since in case of a parsing error, clap simply quit with displaying the error or hint.
Changing `try_parse()` with `parse()` solves the "issue", behaving normally if a "real" parse error is found
2. The `args.rs` module adopts the Clap v3 syntax, since I'm refactoring in order to add the graph subcommand, It's ok if I switch to the new derive syntax?
```rust
#[clap(long, value_parser)]
```
to
```rust
#[arg(long)]
```
And since quoting [Clap changelog ](https://github.com/clap-rs/clap/blob/master/CHANGELOG.md#highlights)

> (derive) Changed the default for arguments from parse to value_parser, removing parse support (#3827, #3981)
    #[clap(value_parser)] and #[clap(action)] are now redundant
I would remove the `[(value_parser)]` too.

3. Since I'm adding the graph subcommand should I change, for coherence, the `--list-jobs` from flag option to a subcommand too?
4. The `--help` flags are disabled in the parser configuration
```rust
#[clap(name="whiz", about, long_about = None, disable_version_flag = true, disable_help_flag = true)]
```
Now that there is more than one subcommand (at least two with version and subgraph) should be better if this is enabled again? 
`whiz help version`
`whiz upgrade --help`
`whiz upgrade -h`
would give the same result

Thanks for your time!

File: src/main.rs [hunk 2]
```diff
@@ -56,18 +55,13 @@ async fn upgrade_check() -> Result<()> {
 }
 
 fn main() -> Result<()> {
-    let args = Args::try_parse()?;
+    let args = Args::parse();
```
Comment (Darioazzali) [inline src/main.rs:58]: Here is the change about the printing error in whiz usage

File: src/main.rs [hunk 3]
```diff
@@ -144,12 +138,36 @@ async fn run(args: Args) -> Result<()> {
         .filter_jobs(&args.run)
         .map_err(|err| anyhow!("argument error: {}", err))?;
 
-    if args.list_jobs {
+    if let Some(Command::ListJobs) = args.command {
         let formatted_list_of_jobs = config.get_formatted_list_of_jobs();
         println!("List of jobs:\n{formatted_list_of_jobs}");
         return Ok(());
```
Comment (Darioazzali) [inline src/main.rs:144]: Reformatted as subcommand

File: src/main.rs [hunk 4]
```diff
@@ -144,12 +138,36 @@ async fn run(args: Args) -> Result<()> {
...
+            .map(|task| graph::Task {
+                name: task.0.to_owned(),
+                depends_on: task.1.depends_on.resolve(),
+            })
+            .collect();
+
+        match graph::draw_graph(filtered_tasks, opts.boxed)
+            .map_err(|err| anyhow!("Error visualizing graph: {}", err))
+        {
+            Result::Ok(..) => {
```
Comment (Darioazzali) [inline src/main.rs:160]: Exit from actix task with stop code 0 on success, 1 on failure reporting the reason

**Comment by zifeo** [comment]
@Darioazzali all good, very good improvement. Could you also update the readme and we are good to go?

**Comment by Darioazzali** [comment]
@zifeo, done!, split the subcommands from flags in the readme cli section

**Review by zifeo — APPROVED** [review]
Thanks for the contribution

**Pull request merged**